### PR TITLE
feat(web): searchable connections switcher

### LINKED ISSUE #396: Replace the connections dropdown with a one that also supports/has search for easier connection switching
Image

### PR DESCRIPTION
**Pull request by jamby77**
Closes #396.

The issue is a screenshot with no text: the connections dropdown, long enough
that finding one means scrolling. This replaces the picker with a combobox that
filters as you type.

## What it does

- **Matches on name *and* `host:port`.** An operator who remembers the port but
  not the label is precisely the case a scrolling dropdown serves worst.
- **Fully keyboard operable.** Arrows move through the **filtered** set, Enter
  selects, Escape closes, focus returns to the trigger. Navigating the
  unfiltered list while a filter is showing would select a row that is not on
  screen, so the highlight resets whenever the query changes.
- **Distinguishes "no connections yet" from "no matches."** Those are different
  problems with different next actions.
- **Shows health per row**, not only on the trigger — a searchable list is the
  natural place to see which connections are up.

## Scope: the picker only

The spec for this issue proposed splitting `ConnectionSelector.tsx` three ways
first. **I deliberately did not.** That file is 1215 lines and also owns the
add-connection flow across three tabs, URL parsing, agent-token generation,
managed-database creation and workspace calls. The user-visible win does not
depend on moving any of it, and a refactor of that size is where this change
would go wrong.

Only the picker — a 17-line block — moved into
`components/connection-selector/`, following the repo convention that
single-use components live in a subfolder named for their parent. The
add-connection flow is untouched, and its existing tests pass unchanged, which
is the evidence that nothing else shifted.

## No new dependency

The spec recommended `cmdk`, the shadcn-idiomatic choice. It is not installed,
and adding a dependency was still an open question — so this builds on
`@radix-ui/react-popover`, already available through the `radix-ui` umbrella the
repo depends on. The cost is hand-written arrow-key handling, which is covered
by tests. Worth revisiting if #397 wants a command palette, where `cmdk` would
earn its place.

## Verification

- 22 tests for the switcher: filtering by name, by host, by port,
  case-insensitivity, selection by id rather than position, arrow navigation
  staying inside the filtered set, Enter selecting nothing when nothing matches,
  the two empty states, the current connection being marked, plus everything the
  review rounds below added
- full `apps/web` suite green: 49 files, 380 tests, including the existing
  `ConnectionSelector.test.tsx` untouched
- `tsc --noEmit` clean; eslint clean on the new component

## Notes for review

Prettier was run only on the new files. `ConnectionSelector.tsx` is already
prettier-dirty on master, so formatting it would have buried a 5-line change in
unrelated reformatting.

`pnpm lint` fails repo-wide on master today (105 problems), and every `*.test.tsx`
file hits a parser error because `tsconfig.json` excludes test files from the
project eslint resolves against. Both are pre-existing and untouched here.

## Review rounds

Every finding verified before acting. Several were regressions from my own
earlier fixes in this PR, which is worth naming rather than smoothing over:

- **The highlight survived close and reopen**, so Enter selected a row the
  operator never chose.
- **Arrow keys never scrolled the highlight into view** — on a height-capped,
  scrollable list, which is the whole case this component exists for.
- **Fixing that broke the keyboard.** Scrolling moves rows under a stationary
  cursor, the browser fires `mouseenter`, and the highlight jumped away from the
  keyboard target. The guard now keys on the pointer actually changing
  coordinates, rather than on any `mousemove` — which a scroll can emit without
  the pointer going anywhere.
- **`refreshConnections` can shrink the list under an open popover**, leaving the
  index past the end so Enter did nothing. The effective index is derived rather
  than corrected in an effect.
- **ArrowDown was clamped but ArrowUp was not**, so after a shrink several
  presses moved nothing.
- **Filtering did not re-scroll**, so an already-scrolled list could leave the
  active row off screen.
- **`aria-selected` marks the connection in use**, not the keyboard position, so
  arrow navigation was invisible to a screen reader. `aria-activedescendant` now
  tracks the highlight.
- **Two of my own tests were wrong**: the scroll assertion was vacuous (opening
  already triggers the effect, so it passed before ArrowDown ran), and a
  `Element.prototype` stub was left installed for every later test in the worker.

## Follow-up

#397 (keybindings) names toggling this component and selecting from it as its
first requirement. It is exported as its own component for that reason; the
remaining piece is an imperative open/focus handle, which belongs in that PR
alongside the shortcut registry.

---

> [!NOTE]
> **Low Risk**
> UI-only change to connection picking in the web app; selection still goes through the existing `setConnection` callback with no backend or auth changes.
> 
> **Overview**
> Replaces the multi-connection **Select** dropdown in `ConnectionSelector` with a new **`ConnectionSwitcher`** combobox so long connection lists can be filtered instead of scrolled.
> 
> The switcher adds a **search field** (case-insensitive match on **name** and **`host:port`**), a **scroll-capped list** with **health dots on every row**, and **keyboard + mouse** selection (arrows move through the **filtered** set, Enter selects, query and highlight reset on close). It uses **Radix Popover** (no new deps) and includes **`aria-activedescendant`** for screen-reader navigation, plus guards for list refresh/shrink while open and for scroll-induced **mouseenter** stealing keyboard highlight.
> 
> **22 unit tests** cover filtering, selection by id, empty states, and the edge cases above; the rest of `ConnectionSelector` (add-connection flow) is unchanged.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 063728f8093b1858ddf48f0ef9f6e16d5eb0c097. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added a searchable connection switcher in place of the previous selector.
  * Filter connections by name or host and port with case-insensitive matching.
  * View connection health indicators and the currently selected connection.
  * Select connections using the keyboard or mouse, with improved navigation, accessibility, and scrolling.
  * Added clear empty-state messaging when no connections or matches are available.
  * Search terms reset when reopening the switcher for a fresh selection experience.
  * Improved popover sizing and text handling for narrow screens.

### CODE REVIEW

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  return (
+    <Popover.Root
+      open={open}
+      onOpenChange={(next) => {
+        setOpen(next);
+        if (!next) {
+          setQuery('');
+        }
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx:107]: Confirmed and fixed in 2c21501e. Closing cleared `query` but left `activeIndex`, so highlighting a later row, dismissing, reopening and pressing Enter selected a row the operator never chose. `activeIndex` now resets alongside `query` in `onOpenChange`, and a regression test walks that exact sequence: arrow down, Escape, reopen, Enter — and asserts the first row is selected.

[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx:107]: Confirmed and fixed in 2c21501e. Closing cleared `query` but left `activeIndex`, so highlighting a later row, dismissing, reopening and pressing Enter selected a row the operator never chose. `activeIndex` now resets alongside `query` in `onOpenChange`, and a regression test walks that exact sequence: arrow down, Escape, reopen, Enter — and asserts the first row is selected.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    setOpen(false);
+    setQuery('');
+  }
+
+  function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
+    if (event.key === 'ArrowDown') {
+      event.preventDefault();
+      setActiveIndex((index) => {
+        return Math.min(index + 1, filtered.length - 1);
+      });
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx]: Fixed in 2c21501e — clamped to `Math.min(index + 1, Math.max(filtered.length - 1, 0))`.

Worth noting it was latent rather than reachable: with the highlight now reset on close, and `filtered[activeIndex]` already `undefined` for an empty set, Enter selected nothing either way. But `-1` is not a state this should ever hold, and it would bite the moment that guard changed.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                    key={connection.id}
+                    type="button"
+                    role="option"
+                    aria-selected={isCurrent}
+                    onMouseEnter={() => setActiveIndex(index)}
+                    onClick={() => handleSelect(connection.id)}
+                    className={cn(
+                      'w-full flex items-center gap-2 rounded-sm px-2 py-1.5 text-sm text-left',
+                      index === activeIndex && 'bg-accent',
+                    )}
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx:213]: Real, and the most user-visible of the three — the list is `max-h-64 overflow-y-auto`, so on a long connection list, which is the case this component exists for, arrow keys could highlight and select rows nobody could see.

Fixed in 2c21501e with a DOM effect that scrolls the active option into view with `block: "nearest"` whenever the highlight or open state changes. `nearest` rather than `center` so it only scrolls when the row is actually outside the viewport, instead of jumping the list on every keypress. Covered by a test that stubs `scrollIntoView` and asserts it fires on arrow-down.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  const filtered = useMemo(() => {
+    return connections.filter((connection) => {
+      return matches(connection, query);
+    });
+  }, [connections, query]);
+
+  useEffect(() => {
+    const active = listRef.current?.querySelectorAll('[role="option"]')[activeIndex];
+    active?.scrollIntoView({ block: 'nearest' });
+  }, [activeIndex, open]);
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx]: Confirmed and fixed in f16c2132 — a real interaction bug, and it shows up on precisely the long lists this switcher was built for.

`mouseenter` is now ignored while the keyboard is driving, and the mouse takes over again on the next real `mousemove` over the list, so pointer users lose nothing. Two tests: one arrows down, fires `mouseenter` on a different row and asserts Enter still selects the keyboard target; the other moves the mouse first and asserts the highlight follows it.

[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx]: Correct — `refreshConnections` can land while the popover is open, and an `activeIndex` left pointing past the end made Enter silently do nothing.

Fixed in f16c2132 by deriving the effective index rather than correcting it after the fact:

    const effectiveIndex = filtered.length === 0 ? -1 : Math.min(activeIndex, filtered.length - 1);

Derived instead of clamped in an effect for two reasons: it cannot be briefly wrong between render and effect, and `react-hooks/set-state-in-effect` is an error in this config. Covered by a rerender test that shrinks the list under an open popover and asserts Enter still selects a visible row.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                // selects a row the operator cannot see.
+                setActiveIndex(0);
+              }}
+              onKeyDown={handleKeyDown}
+              placeholder="Search by name or host:port"
+              className="w-full bg-transparent py-2 text-sm outline-none placeholder:text-muted-foreground"
+            />
+          </div>
+
+          <div ref={listRef} role="listbox" className="max-h-64 overflow-y-auto p-1">
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx]: Right — `aria-selected` marks the connection in use and says nothing about where the keyboard is, so arrow navigation was invisible to a screen reader.

Fixed in f16c2132: each option gets a stable id from `useId`, the searchbox binds `aria-activedescendant` to the highlighted one, and `aria-controls` points at the listbox. A test asserts the attribute matches the highlighted option after ArrowDown, not the current connection.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.test.tsx [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  it('scrolls the highlighted row into view', () => {
+    // The list is height-capped and scrollable, which is the whole point for a
+    // long connection list — a highlight that never scrolls is unusable.
+    const scrollIntoView = vi.fn();
+    Element.prototype.scrollIntoView = scrollIntoView;
+    open();
+
+    fireEvent.keyDown(screen.getByRole('searchbox'), { key: 'ArrowDown' });
+
+    expect(scrollIntoView).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.test.tsx]: Correct, the assertion was vacuous — `open()` triggers the scroll effect, so `toHaveBeenCalled()` passed before ArrowDown ever ran. Fixed in f16c2132 by clearing the spy after `open()` so the assertion proves navigation scrolled.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.test.tsx [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    fireEvent.keyDown(search, { key: 'Enter' });
+
+    expect(onSelect).toHaveBeenCalledWith('a');
+  });
+
+  it('scrolls the highlighted row into view', () => {
+    // The list is height-capped and scrollable, which is the whole point for a
+    // long connection list — a highlight that never scrolls is unusable.
+    const scrollIntoView = vi.fn();
+    Element.prototype.scrollIntoView = scrollIntoView;
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.test.tsx:180]: Correct — assigning to `Element.prototype.scrollIntoView` leaves the stub installed for every later test sharing the worker. Fixed in f16c2132: the original is captured once and restored in `afterEach`.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+          </div>
+
+          <div
+            ref={listRef}
+            id={`${optionIdPrefix}-listbox`}
+            role="listbox"
+            className="max-h-64 overflow-y-auto p-1"
+            onMouseMove={() => {
+              keyboardNav.current = false;
+            }}
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx:185]: Fixed in 9a9f27cd, and it did undo the guard I had just added.

Rather than depend on which browsers emit synthetic mousemove during scroll, the guard now keys on the pointer actually changing position: `mousemove` records `clientX/clientY` and only hands control back when the coordinates differ from the previous event. A scroll-emitted event at an unchanged position no longer counts as intent.

The cost is one event before the mouse takes over, which real movement always provides — a genuine drag emits a stream at changing coordinates. Two tests cover it: repeated moves at identical coordinates keep the keyboard in charge, and moves at changing coordinates hand it back.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+  const activeOptionId =
+    effectiveIndex >= 0 ? `${optionIdPrefix}-option-${effectiveIndex}` : undefined;
+
+  useEffect(() => {
+    if (effectiveIndex < 0) {
+      return;
+    }
+    const active = listRef.current?.querySelectorAll('[role="option"]')[effectiveIndex];
+    active?.scrollIntoView({ block: 'nearest' });
+  }, [effectiveIndex, open]);
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx]: Correct — filtering with the highlight already at index 0 changes neither `effectiveIndex` nor `open`, so the effect never re-ran and a list the user had scrolled kept its offset with the active row potentially out of view.

Fixed in 9a9f27cd by adding `filtered` to the dependencies, with a regression test that clears the scroll spy after opening and asserts filtering scrolls again.

File: apps/web/src/components/connection-selector/ConnectionSwitcher.tsx [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+      event.preventDefault();
+      keyboardNav.current = true;
+      setActiveIndex((index) => {
+        return Math.min(index + 1, Math.max(filtered.length - 1, 0));
+      });
+      return;
+    }
+    if (event.key === 'ArrowUp') {
+      event.preventDefault();
+      keyboardNav.current = true;
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/web/src/components/connection-selector/ConnectionSwitcher.tsx:82]: Real, and an inconsistency I introduced when deriving the clamped index — ArrowDown was rewritten to clamp while ArrowUp still stepped from the raw value, so after a shrink the two diverged and several presses moved nothing visible.

Fixed in 9a9f27cd: both now step from `effectiveIndex`, so the keyboard always moves relative to what is actually highlighted. Covered by a test that shrinks the list under an open popover and asserts the first ArrowUp moves.

**Review by KIvanow — CHANGES_REQUESTED** [review]
Currently we scale the dropdown based  on the width of the elements Image

The new implementation has  a horizontal scroll, which should be avoided either by stretching the width like we  do currently live, or by clipping the name/url  of the connections,  or  propose another alternative. I am  open to other ideas.
Image

Also, in some cases we are skipping the name of the connections - notice the ones in  the middle
Image
Compared with the full name
Image

**Comment by jamby77** [comment]
Both confirmed, and they turned out to be the same bug. Fixed in 063728f8.

`truncate` sets `overflow: hidden`, and a flex item whose overflow is not `visible` has its **automatic minimum size resolve to 0** instead of to its content width. The name span could therefore shrink to nothing, while the `host:port` span next to it was `flex-shrink-0` and could not give up a pixel. On a fixed `w-[var(--radix-popover-trigger-width)]` popover with a hosted URL that is wider than the whole sidebar, the name lost the whole fight:

- the three `driving-gecko-108653.upstash.io:6379` rows rendered their name at **0px** — which is the missing names in your second screenshot, not a data problem; the names are there
- the host still overflowed the container even after the name vanished, and since `overflow-y: auto` forces the computed `overflow-x` to `auto`, that became the horizontal scrollbar in your first screenshot

Two changes:

1. **The popover grows with its rows**, like Manage Connections does — `w-auto`, floored at `max(16rem, var(--radix-popover-trigger-width))` so it never looks narrower than the trigger it hangs off, capped at `min(32rem, var(--radix-popover-content-available-width))` so it cannot run past the viewport edge on a small window.
2. **Both labels are `min-w-0 truncate`** and neither is `flex-shrink-0`, so when a row genuinely cannot fit they ellipsize in proportion to their length — the long hosted URL gives ground before the name does, rather than the name silently losing all of it.

That is your first option with the third as the fallback, since stretching alone still breaks on a viewport narrower than the widest row.

I reproduced both symptoms in isolation with your connection list and measured the fix rather than eyeballing it: on the `Upstash test` row the name renders at **0px before / 78px after**, and the list's horizontal overflow goes to **0**. All 380 web tests pass and `tsc` is clean; I also checked the built CSS actually contains the two arbitrary `max()`/`min()` utilities rather than Tailwind dropping them silently.

Worth knowing for review: none of this is reachable by the existing tests — jsdom has no layout engine, so every one of these assertions would pass on the broken markup too. This one needs eyes.

**Review by KIvanow — APPROVED** [review]
Looks  good now!
